### DISCUSSION TITLE
nlmixr2 not censoring bloq data

### DISCUSSION THREAD

**Discussion by akosuaagyeman** [nlmixr2/nlmixr2#46] [Q&A] [answered]
Hi nlmixr team,

Please, I have previously used nlmixr to censor bloq data by applying the M4 method which works correctly. However, upon using nlmixr2, the M4 censoring is ignored (i.e., DV for CENS=1 remains the same). 
Below are the codes I used with the **theo_md** dataset as example.

Many thanks in advance.

**nlmixr code**

```
rm(list=ls())
###### set wd  ######
mod.dir <- ""
setwd(mod.dir)
library(nlmixr)
library(vpc)
library(patchwork)
##### read data ####
PK.dat <- (theo_md)
#
# creating a column 'CENS' for stratifying BLOQ values (set LOQ to 3).
# Assign CENS = 1 for bloq values, otherwise CENS = 0.
PK.dat$CENS[PK.dat$DV < 3 & PK.dat$AMT == 0] <- 1
PK.dat$CENS[PK.dat$DV >= 3 & PK.dat$AMT == 0] <- 0
#
# Set DV to LOQ for all censored items
PK.dat$DV[PK.dat$CENS == 1] <-  3
#
# for DV values with CENS = 1, the DV values are censored between (LIMIT, DV).
# Here we create the LIMIT column and set to a value close to 0 [i.e.,log(0.001) ] for DV values with CENS = 1. 
PK.dat$LIMIT <- NA
PK.dat$LIMIT[PK.dat$CENS == 1] <- log(0.001)
#
# fit one-compartment PK model
one.cmt <- function() {
  ini({
    tka <- 0.45 # Log Ka
    tcl <- 1 # Log Cl
    tv <- 3.45    # Log V
    eta.ka ~ 0.6
    eta.cl ~ 0.3
    eta.v ~ 0.1
    add.sd <- 0.7
  })
  model({
    ka <- exp(tka + eta.ka)
    cl <- exp(tcl + eta.cl)
    v <- exp(tv + eta.v)
    d/dt(depot) = -ka * depot
    d/dt(center) = ka * depot - cl / v * center
    cp = center / v
    cp ~ add(add.sd)
  })
}

# Check the model
nlmixr(one.cmt)
# Fit the model 
fit.data1 <- nlmixr(one.cmt, PK.dat , est = "focei",
                   table = tableControl(npde = TRUE))
#
# Plot VPC 
fit.data1_vpc <- nlmixr::vpc(fit.data1, show = list(obs_dv = TRUE))
# Extract the simulated data
sim_data <- fit.data1_vpc$sim
# Specify that time is the idv column or it gives an error message asking for idv
sim_data$idv <- sim_data$time
# Then make the vpc with the vpc package:
vpc1 <- vpc::vpc(sim = sim_data,
                 obs = fit.data1,
                 show = list(obs_dv = TRUE),
                 lloq = 3,
                 facet = "rows") 
vpc2 <- vpc::vpc_cens(sim = sim_data,
                      obs = fit.data1,
                      lloq = 3,
                      facet = "rows") 
vpc1/vpc2
# 
# Save VPC
  pdf("fit.data1_vpc.data",width = 6, height = 4)
  vpc1/vpc2
  dev.off()
#
```
![image](https://user-images.githubusercontent.com/88038872/187360385-8838dd10-a265-473d-8f67-be8567942702.png)

**nlmixr2 code**
```
rm(list=ls())
###### set wd  ######
mod.dir <- "/home/akosua/Documents/nlmixr_nlmixr2/"
setwd(mod.dir)
library(nlmixr2)
library(vpc)
library(patchwork)
##### read data ####
PK.dat <- (theo_md)
#
# creating a column 'CENS' for stratifying BLOQ values (set LOQ to 3). 
# Assign CENS = 1 for bloq values, otherwise CENS = 0.
PK.dat$CENS[PK.dat$DV < 3 & PK.dat$AMT == 0] <- 1
PK.dat$CENS[PK.dat$DV >= 3 & PK.dat$AMT == 0] <- 0
#
# Set DV to LOQ for all censored items
PK.dat$DV[PK.dat$CENS == 1] <-  3
#
# for DV values with CENS = 1, the DV values are censored between (LIMIT, DV).
# Here we create the LIMIT column and set to a value close to 0 [i.e.,log(0.001) ] for DV values with CENS = 1. 
PK.dat$LIMIT <- NA
PK.dat$LIMIT[PK.dat$CENS == 1] <- log(0.001)
#
# fit one-compartment PK model
one.cmt <- function() {
  ini({
    tka <- 0.45 # Log Ka
    tcl <- 1 # Log Cl
    tv <- 3.45    # Log V
    eta.ka ~ 0.6
    eta.cl ~ 0.3
    eta.v ~ 0.1
    add.sd <- 0.7
  })
  model({
    ka <- exp(tka + eta.ka)
    cl <- exp(tcl + eta.cl)
    v <- exp(tv + eta.v)
    d/dt(depot) = -ka * depot
    d/dt(center) = ka * depot - cl / v * center
    cp = center / v
    cp ~ add(add.sd)
  })
}

# Check the model
nlmixr2(one.cmt)
# Fit the model 
fit.data2 <- nlmixr2(one.cmt, PK.dat , est = "focei",
                   table = tableControl(npde = TRUE))
#
# Plot VPC 
sim_data <- nlmixr2::vpcSim(fit.data2)
# extract the simulated data
sim_data$dv <- sim_data$sim
# specify that time is the idv column or it gives an error message asking for idv
sim_data$idv <- sim_data$time
# then make the vpc with the vpc package:
vpc1 <- vpc::vpc(sim = sim_data,
                 obs = fit.data2,
                 show = list(obs_dv = TRUE),
                 lloq = 3,
                 facet = "rows")

vpc2 <- vpc::vpc_cens(sim = sim_data,
                      obs = fit.data2,
                      lloq = 3,
                      facet = "rows")

vpc1/vpc2
# 

# Save VPC
  pdf("fit.data2_vpc.data",width = 6, height = 4)
  vpc1/vpc2
  dev.off()
#
```
![image](https://user-images.githubusercontent.com/88038872/187361910-c1c3b734-668f-4509-b0b6-90aaa69c882e.png)

**Comment by mattfidler** [comment, marked as answer]
Hm, there were many fixes to the censoring methods recently.  I'm unsure if they are on CRAN or not.  

I get a quite different view with my current `nlmixr2`:

![image](https://user-images.githubusercontent.com/514778/187578549-e138d71e-1e16-4b9b-8074-eb783f14d93e.png)

To me it seems that the `nlmixr2` that I am using gives a better fit than `nlmixr` and the `nlmixr2` version you have;  This is both in terms of the traditional `vpc` and the censoring `vpc`.  

If you wish to try the development version to see if that is true for you the instructions are below:

```r
remotes::install_github("nlmixr2/nlmixr2data")
remotes::install_github("nlmixr2/lotri")
remotes::install_github("nlmixr2/rxode2")
remotes::install_github("nlmixr2/nlmixr2est")
remotes::install_github("nlmixr2/nlmixr2extra")
remotes::install_github("nlmixr2/nlmixr2plot")
remotes::install_github("nlmixr2/nlmixr2")
```

Note that my version of nlmixr2 calls out the `M4` method *and*  `nlmixr` does not play well with `nlmixr2` and should be run separately.

**Reply by akosuaagyeman** [reply]
Many thanks Matt for the response.
RStudio crash severally after installing the development version but all good now.
The censoring is working now. 
Thanks for the help!

**Reply by mattfidler** [reply]
The Rstudio crash is probably because of a restored environment that contains other versions.  Usually I don't restore the environment so that this doesn't happen.

**Reply by akosuaagyeman** [reply]
Another problem I am encountering now is that, my R scripts which I used nlmixr2 (not the development version) to run previosly is unable to run with the development version and I get the error message below. I have transferred the script to a new file and directory but still encounters the same error message. However, when I switch to the non-development version of nlmixr2,  my previous scripts and the  **theo_md ** script work but censoring is ignored. This is not good for me as I need nlmixr2 to work for my project datafile and not the **theo_md** example.

![image](https://user-images.githubusercontent.com/88038872/188054372-85250e69-8bd7-4e63-88d5-27cdd441cb3c.png)

**Reply by mattfidler** [reply]
This should be fixed in the development version.  I don't have access to your data, so maybe you can check.

**Reply by akosuaagyeman** [reply]
Many thanks Matt. I am now able to get the M4 censoring working with nlmixr2 development version. I had to reinstall R and Rstudio to get it working.
Although the censoring works, the vpc simulations for the censored data is not quite right with nlmixr2. The prediction interval for the simulated censored data spans across the entire proportion. So you can see from below that with the same dataset (**theo_md**), I'm getting different vpc output for the simulated censored data.
Many thanks in advance.

**nlmixr**
![image](https://user-images.githubusercontent.com/88038872/189117500-d494339d-c328-421a-be64-3ab0105edae6.png)

**nlmixr2**
![image](https://user-images.githubusercontent.com/88038872/189117701-9d106d8e-6bce-4b7a-91c0-542bf1e457f0.png)

**Reply by mattfidler** [reply]
Thanks.  Is thes the same code that you were using above?

**Reply by akosuaagyeman** [reply]
Thanks Matt for the reply.
Yes, the above **nlmixr2** plot is from the code above (i.e.,  please see below). This code works perfectly fine for **nlmixr** and was adapted for **nlmixr2** which isn't working properly for **nlmixr2**. I am also using the development version for this.
```
# Plot VPC 
sim_data <- nlmixr2::vpcSim(fit.data2)
# extract the simulated data
sim_data$dv <- sim_data$sim
# specify that time is the idv column or it gives an error message asking for idv
sim_data$idv <- sim_data$time
# then make the vpc with the vpc package:
vpc1 <- vpc::vpc(sim = sim_data,
                 obs = fit.data2,
                 show = list(obs_dv = TRUE),
                 lloq = 3,
                 facet = "rows")

vpc2 <- vpc::vpc_cens(sim = sim_data,
                      obs = fit.data2,
                      lloq = 3,
                      facet = "rows")
```
I have managed to use **vpcPlot()** code documentation which is giving me similar VPC plot as with **nlmixr** with the **theo_md** dataset (please see code and plot below). So this **vpcPlot()** works for **nlmixr2** but not **nlmixr** and the previous code I was using works for **nlmixr** but not **nlmixr2**.
Many thanks again for the support. Very well appreciated.
```
# Plot VPC
vpc1 <- vpcPlot(fit.data2,idv = "time",
                     show = (list(obs_dv=TRUE)),
                     lloq = 3,
                     cens = FALSE)

vpc2 <-vpcCens(fit.data2,idv = "time",
                    lloq = 3.001, 
                    cens = TRUE)

vpc1/vpc2
```
![image](https://user-images.githubusercontent.com/88038872/189817993-11999264-8767-43fb-8a46-534a1f375eaa.png)

**Reply by mattfidler** [reply]
For me, the `vpc::vpc_cens` and `vpcCens` now give the exact same values. I did some updates for this

**Reply by akosuaagyeman** [reply]
Good to know both ways work now. Many thanks for the update.
